turbopack/crates/turbo-tasks-fs/src/glob.rs

```diff
@@ -42,388 +16,622 @@ enum GlobPart {
+
+    // Returns true if the glob might match a filename underneath this `path` where the
+    // path represents a directory.
+    pub fn can_match_in_directory(&self, path: &str) -> bool {
+        debug_assert!(
+            !path.ends_with('/'),
+            "Path should be a directory name and not end with /"
+        );
+        self.directory_match_regex.is_match(path.as_bytes())
     }
 
     pub fn parse(input: &str) -> Result<Glob> {
-        let mut current = input;
-        let mut expression = Vec::new();
-
-        while !current.is_empty() {
-            let (part, remainder) = GlobPart::parse(current, false)
-                .with_context(|| anyhow!("Failed to parse glob {input}"))?;
-            expression.push(part);
-            current = remainder;
+        let tokens: Tokens = Parser::new(input).parse()?;
+        let regex = new_regex(tokens.to_regex().as_str());
+        let directory_match_regex = new_regex(tokens.to_directory_match_regex().as_str());
+
+        Ok(Glob {
+            glob: input.to_string(),
+            regex,
+            directory_match_regex,
+        })
+    }
+}
+
+impl TryFrom<&str> for Glob {
+    type Error = anyhow::Error;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        Glob::parse(value)
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl Glob {
+    #[turbo_tasks::function]
+    pub fn new(glob: RcStr) -> Result<Vc<Self>> {
+        Ok(Self::cell(Glob::parse(glob.as_str())?))
+    }
+
+    #[turbo_tasks::function]
+    pub async fn alternatives(globs: Vec<Vc<Glob>>) -> Result<Vc<Self>> {
+        match globs.len() {
+            0 => Ok(Glob::new("".into())),
+            1 => Ok(globs.into_iter().next().unwrap()),
+            _ => {
+                // Composing the Regexes with a RegexSet would be more efficient since we wouldn't
+                // need to recompile them.
+                let mut new_glob = String::new();
+                new_glob.push('{');
+                for (index, glob) in globs.iter().enumerate() {
+                    if index > 0 {
+                        new_glob.push(',');
+                    }
+                    new_glob.push_str(&glob.await?.glob);
+                }
+                new_glob.push('}');
+                Ok(Glob::new(new_glob.into()))
+            }
         }
+    }
+}
 
-        Ok(Glob { expression })
+fn new_regex(pattern: &str) -> Regex {
+    RegexBuilder::new(pattern)
+        .dot_matches_new_line(true)
+        .build()
+        .expect("A successfully parsed glob should produce a valid regex")
+}
+
+/// The parsing and construction algorithms are based on the the `globset` crate.
+/// After discussing upstreaming our usecase with the authors of `globset`, we decided that a fork
+/// was appropriate given our divergent usecases. See discussion https://github.com/BurntSushi/ripgrep/issues/3049
+/// The original code is dual licensed under the MIT license and the Unlicense.
+/// Copyright (c) 2015 Andrew Gallant
```

Comment:
To keep the licensing more obvious, can we move anything derived from globset into a separate file?

Reply:
Done!